File: backend/src/utils/discovery_parser.py
"""
Parser for LinkedIn DISCOVERY sheet data.
Handles extraction and validation of overall performance metrics.
"""

from datetime import datetime, date
from typing import Dict, Any, Tuple
import re


class DiscoveryData:
    """Overall performance metrics from the DISCOVERY sheet"""
    def __init__(self, start_date: date, end_date: date, total_impressions: int, members_reached: int):
        self.start_date = start_date
        self.end_date = end_date
        self.total_impressions = total_impressions
        self.members_reached = members_reached

    def to_dict(self):
        return {
            "start_date": self.start_date.isoformat(),
            "end_date": self.end_date.isoformat(),
            "total_impressions": self.total_impressions,
            "members_reached": self.members_reached
        }


def parse_date_range(date_range_str: str) -> Tuple[date, date]:
    """
    Parse a date range string in format "MM/DD/YYYY - MM/DD/YYYY"

    Args:
        date_range_str: Date range string from Excel

    Returns:
        Tuple of (start_date, end_date)

    Raises:
        ValueError: If date range format is invalid
    """
    try:
        # Split on " - " (with spaces)
        dates = date_range_str.strip().split(" - ")
        if len(dates) != 2:
            raise ValueError(f"Expected format 'MM/DD/YYYY - MM/DD/YYYY', got '{date_range_str}'")

        start_date = datetime.strptime(dates[0].strip(), "%m/%d/%Y").date()
        end_date = datetime.strptime(dates[1].strip(), "%m/%d/%Y").date()

        return start_date, end_date
    except (ValueError, AttributeError) as e:
        raise ValueError(f"Failed to parse date range '{date_range_str}': {str(e)}")


def parse_integer_value(value: Any) -> int:
    """
    Parse an integer value from Excel cell (could be float, string, or int)

    Args:
        value: Value from Excel cell

    Returns:
        Parsed integer value

    Raises:
        ValueError: If value cannot be converted to int
    """
    try:
        if isinstance(value, str):
            # Remove commas and whitespace
            value = value.replace(",", "").strip()

        return int(float(value))
    except (ValueError, TypeError) as e:
        raise ValueError(f"Failed to parse integer value '{value}': {str(e)}")


def extract_discovery_data(discovery_cells: Dict[str, Any]) -> DiscoveryData:
    """
    Extract discovery data from the 4 cells in the DISCOVERY sheet.

    Expected structure:
    {
        "Overall Performance": "11/11/2024 - 11/10/2025",
        "Impressions": 857000.0,
        "Members reached": 297771.0
    }

    Args:
        discovery_cells: Dictionary with discovery data keys and values

    Returns:
        DiscoveryData object

    Raises:
        ValueError: If required fields are missing or invalid
    """
    required_fields = {"Overall Performance", "Impressions", "Members reached"}

    if not all(field in discovery_cells for field in required_fields):
        missing = required_fields - set(discovery_cells.keys())
        raise ValueError(f"Missing required discovery fields: {missing}")

    try:
        # Parse date range
        date_range_str = str(discovery_cells["Overall Performance"])
        start_date, end_date = parse_date_range(date_range_str)

        # Parse integer values
        impressions = parse_integer_value(discovery_cells["Impressions"])
        members_reached = parse_integer_value(discovery_cells["Members reached"])

        return DiscoveryData(
            start_date=start_date,
            end_date=end_date,
            total_impressions=impressions,
            members_reached=members_reached
        )
    except ValueError as e:
        raise ValueError(f"Failed to extract discovery data: {str(e)}")

File: site/src/utils/excel/discoveryParser.ts
/**
 * Discovery Sheet Parser
 * =====================
 * Parses the "DISCOVERY" sheet from LinkedIn analytics Excel export.
 *
 * The Discovery sheet has a fixed structure with 2 columns (A: Label, B: Value):
 *   Row 1: "Overall Performance" | (empty)
 *   Row 2: "MM/DD/YYYY - MM/DD/YYYY" | (empty)
 *   Row 3: "Impressions" | numeric value
 *   Row 4: "Members reached" | numeric value
 *
 * This parser directly reads cells from known positions for maximum efficiency.
 */
import type { WorkBook } from 'xlsx';
import type { DiscoveryData } from '@utils/excel/types';
import { 
  getCellValue, 
  parseDate, 
  parseNumber, 
  findSheet
} from '@utils/excel/utils';
import { SHEET_NAMES } from '@utils/excel/constants';

/**
 * Parse the Discovery sheet to extract overall performance metrics
 * Sheet has fixed structure with all data in column B:
 *   Row 1, Col B: Date range (format: "MM/DD/YYYY - MM/DD/YYYY")
 *   Row 2, Col B: Impressions (numeric value)
 *   Row 3, Col B: Members reached (numeric value)
 * 
 * @param workbook - Parsed Excel workbook from xlsx library
 * @returns DiscoveryData object with performance metrics, or undefined if sheet not found
 */
export function parseDiscovery(workbook: WorkBook): DiscoveryData | undefined {
  const sheet = findSheet(workbook, SHEET_NAMES.DISCOVERY);
  if (!sheet) {
    console.warn('DISCOVERY sheet not found');
    return undefined;
  }

  try {
    // Read date range from B1 (format: "MM/DD/YYYY - MM/DD/YYYY")
    const dateRangeCell = getCellValue(sheet, 'B1');
    const [startStr, endStr] = String(dateRangeCell || '').split('-').map((s: string) => s.trim());
    
    // Read impressions from B2
    const impressionsValue = getCellValue(sheet, 'B2');
    
    // Read members reached from B3
    const membersValue = getCellValue(sheet, 'B3');

    const start_date = parseDate(startStr);
    const end_date = parseDate(endStr);
    const total_impressions = parseNumber(impressionsValue);

    const discoveryData: DiscoveryData = {
      start_date,
      end_date,
      total_impressions,
      members_reached: parseNumber(membersValue),
    };

    // Calculate average impressions per day if we have valid dates
    if (start_date && end_date && total_impressions > 0) {
      const daysDiff = calculateDateRangeDays(start_date, end_date);
      discoveryData.average_impressions_per_day = Math.round(total_impressions / daysDiff);
    }

    return discoveryData;
  } catch (error) {
    console.error('Error parsing DISCOVERY sheet:', error);
    return undefined;
  }
}

/**
 * Calculate the number of days in a date range (inclusive)
 * @param startDate - ISO date string
 * @param endDate - ISO date string
 * @returns Number of days (minimum 1)
 */
function calculateDateRangeDays(startDate: string, endDate: string): number {
  const start = new Date(startDate);
  const end = new Date(endDate);
  const daysDiff = Math.floor((end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24)) + 1;
  return Math.max(1, daysDiff);
}
